File: 35-words-to-fig/wd2fig.py
# Python program to convert given number in words to integer
# Created by M.V.Harish Kumar on 14-01-2025

def wd2fig(words, num2Wd={}):
    if not num2Wd:
        onesTo20 = ('zero', 'one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine', 'ten',
                    'eleven', 'twelve', 'thirteen', 'fourteen', 'fifteen', 'sixteen', 'seventeen', 
                    'eighteen', 'nineteen')
        tens = ('zero', 'ten', 'twenty', 'thirty', 'forty', 'fifty', 'sixty', 'seventy', 'eighty', 'ninety')
        delims = ('hundred', 'thousand', 'million', 'billion')
        num2Wd.update({wd: (1,i) for i, wd in enumerate(onesTo20)})
        num2Wd.update({wd: (1, i*10) for i, wd in enumerate(tens)})
        num2Wd.update({wd: (10 ** (2 if i == 0 else i*3), 0) for i, wd in enumerate(delims)})

    cur = res = 0
    if any(words.startswith(d) for d in delims):
        words = 'one ' + words
    for word in words.lower().replace(' and ', ' ').split():
        if word not in num2Wd:
            raise ValueError("Illegal word encountered!")

        scale, incr = num2Wd[word]
        cur = cur*scale + incr
        if scale > 100:
            res += cur
            cur = 0

    return res+cur

num = input("Enter a number(in words): ")
print(f"{num!r} in figures: {wd2fig(num)}")

File: 35-words-to-fig/wd2fig.go
// Go program to convert given number in words to integer
// Written by M.V.Harish Kumar on 18-06-2025
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strings"
)

var onesTo20 = [20]string{"zero", "one", "two", "three", "four", "five", "six", "seven", "eight",
		"nine", "ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen",
		"seventeen", "eighteen", "nineteen"}
var tens = [10]string{"zero", "ten", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"}
var delims = [4]string{"hundred", "thousand", "million", "billion"}

func wd2fig(words string, num2wd map[string][2]float64) int {
	var cur, res float64
	cur, res = 0, 0
	for _, d := range delims {
		if strings.HasPrefix(words, d) {
			words = "one " + words
		}
	}
	
	wordList := strings.FieldsSeq(strings.ReplaceAll(strings.ToLower(words), " and ", " "))
	for word := range wordList {
		value, isPresent := num2wd[word]
		if !isPresent {
			panic("Illegal word encountered")
		}
		scale, incr := value[0], value[1]
		cur = cur*scale + incr
		if scale > 100 {
			res += cur
			cur = 0
		}
	}
	return int(res+cur)
}

func main() {
	num2wd := make(map[string][2]float64)

	for i, v := range onesTo20 {
		num2wd[v] = [2]float64{1, float64(i)}
	}

	for i, v := range tens {
		num2wd[v] = [2]float64{1, float64(i*10)}
	}

	for i, v := range delims {
		var expt float64 = math.Pow10(2)
		if i != 0 {
			expt = math.Pow10(i*3)
		}
		num2wd[v] = [2]float64{expt, 0}
	}

	scnr := bufio.NewScanner(os.Stdin)
	fmt.Print("Enter number(in words): ")
	scnr.Scan()
	num := scnr.Text()
	fmt.Printf("%#s in figures: %d\n", num, wd2fig(num, num2wd))
}
